README.md
#swagger-json-output

Used with [`swagger-node-runner`][1], this is a cherry picked pipe fitting to handle
yielded errors and/or yielded results to JSON.


This fitting is useful only for projects that are using [`swagger-node-runner`][1]
version `0.7` or above.


### Important

Generally, the fitting sets for [`swagger-node-runner`][1] the `ctx.statusCode`
and `ctx.output` based on `ctx.error` and `ctx.output` that it is provided with,
which in turn are used by [`swagger-node-runner`][1] it to emit the response, as 
seen at the [`_finish` handler][2] of the underlying `connect_middleware`.


This pipe works with `error`, `statusCode`, and `output` that it finds on the 
context that it's given.
As a result, it cannot work with data passed to `response.write` nor to 
`response.writeHead`.

**In order for the fitting to work as expected user controllers should 
communicate with the runner using the `context` object and the value yielded to
the `next` callback, and should not use not use the response object to write to
the response or response headers.**

If you cannot work this way, this fitting is not for you :(

We recommend to work with [`controllerInterface: pipe`][3].


**NOTE:** there are two cases which cannot be handled by this pipe fitting:
 - 405 error, when user tries on a path that is found in openapi-spec doc an
   HTTP verb that is not found in the doc.
 - 405 error, when the openapi-spec doc points to a non-existing pipe (value 
   in `x-swagger-pipe` attribute - may be derrived by the default-pipe).
   
In both these cases the `swagger-node-runner` infra rejects the request before
it enters the pipe. This package works as a fitting, and runs only as part of
a pipe. 

Except for these 2 cases - it will do the trick :)

## installing
```bash
npm install swagger-json-output --save
```

## Using it as an `onError` handler

Pass it as `onError` handler to your main pipe.

If you're using the template created by `swagger` cli:

1. Find the definition of your main pipe
2. replace
```yaml
      - onError: json_error_handler
```      
with
```yaml
      - onError: swagger-json-output
```      

## Using it as output JSON formatter
1. add this fitting as a last step to your main pipe:
```yaml
      - swagger-json-output
```

## How does it work

Whenever an error is thrown in any of the fitting in that pipe, `bagpipes` 
captures the error as `context.error` and passes the cotnext to this 
fitting.

Regardless to errors, user controllers are expected to pass the response body 
as 2nd argument to the `next` callback. This value is passed by `bagpipes` to 
this fitting as `context.output`.

As this fitting finishes, wether by error or not - the context has:
 - a defined statusCode
 - a defined response content-type, matching `Accept` http header **and** the 
   `produces` defined on the operation.
 - a well-formatted body (currently guaranteed only for application/json)


Flow:
 - **In case of error**, make sure statusCode is escalated to `>= 400` by using the first code
   that is indeed set and is `>= 400`:
    - `context.statusCode` 
    - `context.response.statusCode`
    - `context.error.statusCode`
    - use 500 as default error code
 - assure that the `content-type` of the response matches the content-type
   defined by the `produces` section of the `openapi-spec` of the processed
   operation.
 - ** In case of error**: 
    - creates `context.output` as a serializable clone of the error, making
      sure the clone will include the `err.message` and if  `includeErrStack` 
      is truthful - `err.stack` as well, together with any enumerable property
      that the error is decorated with.
 - if `ctx._preOutput` handler is found - execute it, and pass it the context as argument.
   The handler is executed synchronously (no callback involved).
 - if the `content-type` of the response should be JSON - it formats the  
   output as JSON.
   Whenever the serialization fails 
    - the response code escalates to 500
      response body will be a beautified JSON object which includes:
        - `message`: `unable to stringify body properly`,
        - `stringifyErr`: the stringification error message
        - `bodyInspect`: Array lines resulted by `util.inspect`ing the `body`.

       
## Configuring the fitting 
Supported options:
 - `includeErrStack` - whenever truthful, in case of error is thrown or yielded
    by fittings in the pipe, the error stack is included in responses.
 - `beautifyJson` - meant for Dev/Integration envs, where you want to `curl` your API and just read.
   The error-stack is optimized for this beautification.

1. Create a fitting definition before your pipe
2. Use the configured fitting instead the raw fitting name. 

Example:

in `default.yaml`
```yaml
  bagpipes: 
    _output:
      name:                   swagger-json-output
      beautifyJson:           false
      includeErrStack:        false

    _router:
      name:                   swagger_router
      controllersInterface:   pipe

    swagger_controllers:
      - onError:              _output
      - swagger_cors
      - swagger_params_parser
      - swagger_security
      - swagger_validate
      - express_compatibility
      - _router
      - _output
``` 

in `dev.yaml`
```yaml
  bagpipes: 
    _output:
      beautifyJson:           true
      includeErrStack:        true
```      

## Last minute modifications to output

If you need to perform a last-minute modification to the output, you can 
provide a synchronous handler and place it on the context as `ctx._preOutput`.

The function is provided two arguments
  - reference to error from which body is created (or null, if body is not 
    result of an error)
  - the context itself, so you don't have to use the keyword `this`.
  
Example:

```javascript
module.exports = function(fittingDef) {
  return function(ctx, next) {
      ctx._preOutput = lastMomentModifyCtx;
      next()
  }
}

function lastMomentModifyCtx(err, ctx) {
    //err        - if the output is a result of an error - this will be
    //             the original error that was formatted into the body
    //ctx.output - will contain the output, or the output created 
    //             from a thrown/yielded error
}
```

The usecase that brought this feature is a proprietary fitting that executes 
early in the pipeline, collects tools (di) and prepares an envelope response,
where by corporate rules any reponse provided by any step must be contained in
this envelope.

So, in fact, the fitting does all the di and prepare the envelope before 
all user-code parts (mainly security-handlers and router controllers), gathers
data to this envelope as execution of the request progresses, and uses the hook 
to enrich and contain the response using the `ctx._preOutput` hook.
 
## Future
 - design handling of multiple content-types

## Contribute
 - Using PRs :).
   If you intend to add functionality - please discuss it with us first on an 
   issue - just to help maintain the spirit of the project :)
 - make sure all tests pass
 
 Thanks!

## Lisence

MIT, and that's it :)
 
[1]: https://www.npmjs.com/package/swagger-node-runner
[2]: https://github.com/theganyo/swagger-node-runner/blob/master/lib/connect_middleware.js#L68
[3]: https://github.com/theganyo/swagger-node-runner/wiki/Controllers-Interface


CHANGELOG.md
swagger-result
================

## 1.1.3
 - fixed errorToOutputIfAny, function return error and provide to preOutput

## 1.1.1, 1.1.2
 - fixes some incorrect/misleading info on README.md. 
   No code changes involved.

## 1.1.0
 - support ctx._preOutput - additional setting and formatting for context
 
## 1.0.0
 - first release 
 
## 0.1.0
 - Initial commit




test/lib.test.js
process.env.TEST_MODE = true;
var sut     = require('../');
var request = require('mocha-ui-exports-request');
var e2e     = require('./util/e2e');

module.exports = { 
  "swagger-result" : {

    "should be a factory function that names 2 arguments - fittingDef, bagpipes" : function() {
        Should(sut)
          .be.a.Function()
          .have.property("length", 2)
    },
    "when used with fittingDef and bagpipes" : {
      "should return a startegy function that names 2 arguments - ctx, next" :  function() {
          Should(sut({},{}))
            .be.a.Function()
            .have.property("length", 2)            
      },
    },
    "when named as last step on controller pipe for operation that produces `application/json`" : {
      "and controller leaves ctx.output that cannot be serialized" : serializationErrorSuire({}),
      "and controller leaves ctx.output that serializes successfully" : serializationOkSuite({
        path: "/hello"
      })
    },
    "when named as onError pipe handler for operations that produce `application/json`" : {
      "and a fitting in the pipe leaves error on context" : errorCaseSuite({
        path: "/operation-leaves-error"
      }),
      "and a fitting in the pipe throws error" : errorCaseSuite({
        path: "/operation-throws"
      }),
      "and a fitting in the pipe yields an error" : errorCaseSuite({
        path: "/operation-yields-error"
      }),
      "and no error is yielded, thrown or left on cotnext" : serializationOkSuite({}),
      "and content-type left on the response does not match the content-type declared by produces" : {
        "should correct the content-type" : null,
        "and a logger is found on the request" : {
          "should emit a warning to the requet logger" : null
        }
      }
    }
  }
};

function errorCaseSuite(ctx) {
    
    return {
      beforeAll: function(done) {
          //TODO - run setup
          //ctx.results = {
          //  statusCode: 
          // }
          done()
      },
      "and fittingDef.beautifyJson is falsful" : {
        "should leave the error to the framework's default handler"  : null
      },
      "and fittingDef.beautifyJson is truthful" : {
        "and context.statusCode is set to a code >= 400" : {
          "response statusCode should be the ctx.statusCode" : null
        },
        "and context.statusCode is NOT set to a code >= 400" : {
          "and response.statusCode is set to a code >= 400" : {
            "response statusCode should be the value on response.statusCode" : null
          },
          "and response.statusCode is NOT set to a code >= 400" : {
            "and error object has .statusCode" : {
              "response statusCode should be the value on error.statusCode" : null
            }, 
            "and error.statusCode is NOT set to a code >= 400" : {
              "response statusCode should be 500" : null
            }
          }
        },
        "and error object is serializable" : {
          "response.statusCode should be preserved as found by statusCode logic" : null,
          "response.body" : {
            "should have .message as the error message" : null,
            "should have .stack as error.stack split by new-line": null,
            "should include any enumerable attribute error was decorated with": null
          }
        },
        "and error object is not serializable" : serializationErrorSuire(ctx)
      }
    }
}

function serializationErrorSuire(ctx) {
    return {
      beforeAll: function(done) {
          
          done()
      },
      "response.statusCode should be 500" : null,
      "response.body" : {
        "should have .message : unable to stringify body properly" : null,
        "should have .stringifyErr as the message of stringification error" : null,
        "should have .bodyInspect as util.inspect of the error, split by new-line" : null
      }
    }
}

function serializationOkSuite(ctx) { 
    return {
      beforeAll: function(done) {

        done()
      },      
      "statusCode should be as meant by controller method" : null,
      "body should be as passed by controller method" : null
    }
}

test/util/e2e.js
var extend   = require('util')._extend;
var fs       = require('fs');
var fork     = require('child_process').fork;
var debug    = require('debug')('util/e2e');
var log;

var defaults = {
    logPath      : "./e2e.log",
    readyNotice  : "listening on port",
    console      : console,
    args         : ["./node_modules/istanbul/lib/cli", "cover", "--dir", "./coverage/e2e-test"],
    timeout      : 10000,
    slow         : 5000
};

e2e.setupFixture = setupFixture;

module.exports = e2e;

function e2e(options) {
    ctx = extend( extend({}, defaults), options );

    ctx.args.push(options.svc);

    e2e.tearDown = function ctx_teardown(done) {
        tearDown(ctx, done)
    };
    
    e2e.setupFixture = function ctx_setupFixture(fixture, done) {
        setupFixture(ctx, fixture, done)
    }

    return function ctx_setup(done) {
        setup(ctx, this, done)
    }
}


/**
 @param {object} options
 @param {string} options.sut - system under test - path to the script that
 runs the target server
 @param {string} options.logPath - path to log file
 @param {string} options.readyNotice - output line expected on stdout of
 the started target service that indicates that the service is running
 @param {mocha.Test} test - the test context that implements .timetout(n), .slow(n) ....
 @param {callback} callback
 */
function setup(ctx, test, done) {

    try { fs.unlinkSync( ctx.logPath ) } catch(e) {}

    test.timeout(ctx.timeout);
    test.slow(ctx.slow);

    var log =
        ctx.log =
            fs.createWriteStream(ctx.logPath, { flags: "a"} );

    log.writable = true;

    var child =
        ctx.child =
            fork(ctx.args.shift(), ctx.args, {
              env: process.env, 
              stdio: ['pipe', 'pipe', 'pipe', 'ipc'] 
            });

    child.stderr.on('data', function(data) {
        data = data + "";
        if (log.writable) log.write("ERR: " + data);
        if (~data.indexOf("Error: listen EADDRINUSE")) {
            done(new Error("    >> Server could not start: Address In Use"));
        }
    });
    child.stdout.on('data', function(data) {
        data = data.toString();
        if (log.writable) log.write(data);

        if (~data.indexOf(ctx.readyNotice)) {
            ctx.console.log("    >> service started: %s", ctx.args.join(' '));
            done();
        }
    });

    child.on('exit', function(e) {
        child.exitted = true;
        ctx.console.log("    >> service termination ended", e || "OK");
        ctx.log.writable = false;
        ctx.log.end(function() {
            child.emit('--over--')
        });
    });
}

/**
 @param {context} ctx
 @param {callback} callback
 */
function tearDown(ctx, done) {
    if (ctx.child.exitted) return done()
    ctx.child.on('--over--', done);
    ctx.child.send('die');
}

/**
 @param {context} ctx
 @param {callback} callback
 @returns {Promise}
 */
function setupFixture(ctx, fixture, next) {
    ctx.child.once('message', messageApproval);
    debug("sending fixture %j", fixture)        
    ctx.child.send(fixture);
        
    function messageApproval(ok) {
    debug("fixture sent", ok)        
      next( ok === "ok" 
        ? null
        : new Error("failed to setup server")
      )
    }
}


lib/index.js
var util = require('util');
var merge = require('lodash/merge');
var debug = require('debug')('swagger:swagger-result');
var xJson = /json/;

exports = module.exports = openapi_result_fcty;

if (process.env.TEST_MODE) {
    openapi_result_fcty.assureStatusCode = assureStatusCode;
    openapi_result_fcty.assureCorrectResponseContentType = assureCorrectResponseContentType;
    openapi_result_fcty.errorToOutputIfAny = errorToOutputIfAny;
    openapi_result_fcty.preOutput = preOutput;
    openapi_result_fcty.beautifyJsonContents = beautifyJsonContents;
}

function openapi_result_fcty(fittingDef, bagpipes) {
    debug('created');
    //TRICKY: prefer one load-time-if over if-per-request
    return fittingDef.beautifyJson
        ? function openapi_result(ctx, next) {
        debug('exec');

        assureStatusCode(ctx);
        assureCorrectResponseContentType(ctx);
        var err = errorToOutputIfAny(ctx, fittingDef);

        preOutput(err, ctx);
        beautifyJsonContents(ctx, fittingDef);

        next(null, ctx.output);
    }
        : function openapi_result(ctx, next) {
        debug('exec');

        assureStatusCode(ctx);
        assureCorrectResponseContentType(ctx);
        var err = errorToOutputIfAny(ctx, fittingDef);
        preOutput(err, ctx);

        next(null, ctx.output);
    };
}

// use special client settings
function preOutput(err, ctx) {
    if (!ctx._preOutput)
        return;

    try {
        ctx._preOutput(err, ctx);
    } catch (e) {
        debug("unable use preOutput", e.message);
        ctx.statusCode = 500;
        ctx.output = {
            message: "error invoking _preOutput handler",
            error: e.message,
            stack: e.stack
        }
    }
}

function assureStatusCode(ctx) {
    var err = ctx.error;
    var res = ctx.response;

    if (err) { //i.e - error code must be above 400
        //escalate statusCode. 
        // First higher than 400 is used
        //   1. - ctx.statusCode (left for us by pipe-fitting or user controller)
        //   2. - res.statusCode (left for us by other low level mw)
        //   3. - ctx.error.statusCode/status (error thrown with statusCode/status)
        // Else - use default: 500
        switch (true) {
            case ctx.statusCode >= 400:
                //cool. we're there
                return;

            case res.statusCode >= 400:
                ctx.statusCode = res.statusCode;
                return;

            case statusIn(err) >= 400:
                ctx.statusCode = err.statusCode || err.status;
                return;

            default:
                ctx.statusCode = 500;
        }
    }

    if (!ctx.statusCode) ctx.statusCode = 200;

    function statusIn(err) {
        //TRICKY: sails is different - it uses .status instead of .statusCode
        return err.statusCode || err.status
    }
}

function errorToOutputIfAny(ctx, fittingDef) {
    var err = ctx.error;

    if (!err) return;

    delete ctx.error;

    if (err.statusCode) {
        ctx.statusCode = err.statusCode;
        delete err.statusCode;
    }

    //decouple error view from Error object
    //and make sure stack and message enumerate
    ctx.output = merge({
        message: err.message,
        stack: fittingDef.includeErrStack && err.stack && err.stack.split("\n") || undefined
    }, err);

    return err;

}

function assureCorrectResponseContentType(ctx) {
    var operation = ctx.request.swagger.operation;
    var produces = operation.produces;
    var contentType = ctx.headers["content-type"];
    if (contentType
        && ~produces.indexOf(contentType)
    )
        return ctx.response.setHeader("content-type", contentType);

    if (contentType) warnAboutBadContentType("context");

    contentType = ctx.response.getHeader("content-type");
    if (contentType
        && ~produces.indexOf(contentType)
    )
        return ctx.headers["content-type"] = contentType;

    if (contentType) warnAboutBadContentType("response.getHeader");

    contentType = ctx.request.headers.accept;

    if (!contentType || "*/*" == contentType)
        contentType = produces[0];

    if (contentType) {
        ctx.headers["content-type"] = contentType;
        ctx.response.setHeader("content-type", contentType);
    }

    function warnAboutBadContentType(at) {
        var log = ctx.log || ctx.request.log || console;

        (log.warn || log.error)({
            operationPath: operation.path,
            operationProduces: produces,
            foundType: contentType,
            foundAt: at,
        }, "auto correcting content-type: [%s] does not match produces [%s] of operation: [%s]", contentType, produces, operation.path);
    }
}

function beautifyJsonContents(ctx, fittingDef) {
    var type = ctx.headers["content-type"];
    if (!xJson.test(type))
        return debug("not beautified", type);

    debug("exec", type);

    try {
        ctx.output = JSON.stringify(ctx.output, null, 2)
    } catch (ex) {
        debug("unable to stringify body", ctx.output);
        ctx.statusCode = 500;
        ctx.output = JSON.stringify({
            message: "unable to stringify body properly",
            stringifyErr: ex.message,
            inspect: util.inspect(ctx.output, {depth: 10, breakLength: 1})
                .split("\n")
        }, null, 2);
    }
}
